src/net.cpp

```diff
@@ -2820,40 +2788,124 @@ bool CConnman::NodeFullyConnected(const CNode* pnode)
-        if (nMessageSize)
-            pnode->vSendMsg.push_back(std::move(msg.data));
+        //form the AAED (encipher and append tag)
+        if (!pnode->m_encryption_handler->EncryptAppendMAC(serialized_envelope)) {
+            LogPrintf("Encryption failed, peer=%d\n", pnode->GetId());
+            pnode->fDisconnect = true;
+            return;
+        }
+        {
+            LOCK(pnode->cs_vSend);
+            bool optimisticSend(pnode->vSendMsg.empty());
 
-        // If write queue empty, attempt "optimistic write"
-        if (optimisticSend == true)
-            nBytesSent = SocketSendData(pnode);
+            //log total amount of bytes per command
+            pnode->mapSendBytesPerMsgCmd[msg.command] += nTotalSize;
+            pnode->nSendSize += nTotalSize;
+
+            if (pnode->nSendSize > nSendBufferMaxSize)
+                pnode->fPauseSend = true;
+            pnode->vSendMsg.push_back(std::move(serialized_envelope));
+
+            // If write queue empty, attempt "optimistic write"
+            if (optimisticSend == true)
+                nBytesSent = SocketSendData(pnode);
+        }
+    } else {
+        std::vector<unsigned char> serializedHeader;
+        serializedHeader.reserve(CMessageHeader::HEADER_SIZE);
+        uint256 hash = Hash(msg.data.data(), msg.data.data() + nMessageSize);
+        CMessageHeader hdr(Params().MessageStart(), msg.command.c_str(), nMessageSize);
+        memcpy(hdr.pchChecksum, hash.begin(), CMessageHeader::CHECKSUM_SIZE);
+
+        CVectorWriter{SER_NETWORK, INIT_PROTO_VERSION, serializedHeader, 0, hdr};
+        {
+            LOCK(pnode->cs_vSend);
+            bool optimisticSend(pnode->vSendMsg.empty());
+
+            //log total amount of bytes per command
+            pnode->mapSendBytesPerMsgCmd[msg.command] += nTotalSize;
+            pnode->nSendSize += nTotalSize;
+
+            if (pnode->nSendSize > nSendBufferMaxSize)
+                pnode->fPauseSend = true;
+            pnode->vSendMsg.push_back(std::move(serializedHeader));
+            if (nMessageSize)
+                pnode->vSendMsg.push_back(std::move(msg.data));
+
+            // If write queue empty, attempt "optimistic write"
+            if (optimisticSend == true)
+                nBytesSent = SocketSendData(pnode);
+        }
     }
     if (nBytesSent)
         RecordBytesSent(nBytesSent);
 }
 
+void CConnman::SendStealthEncryptionHandshakeData(CNode* pnode)
+{
+    // initialize encryption, generate ephemeral key
+    assert(pnode->m_encryption_handler == nullptr);
+    pnode->m_encryption_handler = std::make_shared<BIP151Encryption>();
+
+    // get encryption handshake data
+    std::vector<unsigned char> handshake_data;
+    pnode->m_encryption_handler->GetHandshakeRequestData(handshake_data);
+
+    // push handshake data
+    LogPrint(BCLog::NET, "Send stealth encryption handshake payload of %d bytes, peer=%d\n", handshake_data.size(), pnode->GetId());
```

Comment:
You're using the term "stealth encryption" in various, but this term isn't used in the BIP.